_posts/2023-01-11-no-builder-in-kotlin.markdown

```diff
@@ -0,0 +1,184 @@
+After all your organisation might give out usernames like "1234" and allow passwords like "a_schmidt". With Lombok's help, this would look like
+
+```java
+@Value
+class Username {
+    private String value;
+}
+
+@Value
+class Password {
+    private String value;
+}
+
+@Data
+class A {
+    private Role role;
+    private String x1;
+    private Username username;
+    private Password password;
+    private String x2;
+    private String x3;
+}
+```
+
+Now even invoking the constructor without a Builder mixed up would fail (at least for username and password), since
+
+```java
+var a = new A(null, null, new Password("1234"), new Username("a_schmidt"), null, null);
+```
+is a compiletime error.
+
+## Why this isn't a problem in Kotlin to begin with
+
+You don't need to employ the Builder in Kotlin because Kotlin provides optional parameters, named parameters and nullability also plays a certain part. 
+So given a class declaration like this:
+
+```kotlin
+data class A(
+    val role: Role = Role.User, // provide a sensible default if such a default should exist 
+    val x1: String? = null,     // or use a nullable type and initialise it with null by default
+    val username: String,
+    val password: String,
+    val x2: String? = null,
+    val x3: String? = null,
+)
+```
+
+you can invoke the constructor like this
+
+```kotlin
+val a = A(
+    username = "a_schmidt",
+    password = "1234",
+)
+```
+
+As you can see, this looks very close to what a Builder in Java gives you, but comes out of the box in Kotlin.
+Yes, you do have to think a bit more when writing the constructor in order to determine which parameters are optional and which ones have to be provided every time.
+This gives you additional security though (not even thinking about nullability here). In Java you can write code like
+
+```java
+var a = A.builder()
+            .setPassword("12324")
+            .build();
+```
+
+and create an instance that is bound to trigger a NullPointer exception (or fail a check) at runtime.
+In Kotlin the equivalent code
+
+```kotlin
+val a = A(
+    password = "1234",
+)
+```
+won't even compile. Most likely you will notice it immediately because your IDE will point it out to you.
+
+**Misconception: I heard there was a Lombok plugin for Kotlin. So can't I simply use the @Builder on a Kotlin class anyway?**
+Yes and No.
+Yes, there is a [Lombok plugin for Kotlin](https://kotlinlang.org/docs/lombok.html) with support for the most common annotations.
+Yes, since Kotlin 1.8 there is also support for the @Builder annotation.
+No, you can't use it on your Kotlin class.
+The plugin only allows Kotlin code to understand Java code annotated with Lombok's most common annotations, but it won't let you annotate Kotlin classes with it.
+In more words: It simply helps your Kotlin compiler to understand that a `@Builder` annotation on a java class means
```

Comment:
in more words -> in other words?
ich nehme an du willst hier tatsächlich eine ausführlichere ausführung hinweisen. geläufiger wäre beispielsweise "to be more precise".